nexus/src/app/sagas/instance_create.rs

```diff
@@ -299,11 +311,255 @@ impl NexusSaga for SagaInstanceCreate {
+        }
         builder.append(instance_ensure_action());
         Ok(builder.build()?)
     }
 }
 
+async fn sic_add_network_config(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let net_params = sagactx.saga_params::<NetworkConfigParams>()?;
+    let which = net_params.which;
+    let instance_id = net_params.instance_id;
+    let params = net_params.saga_params;
+    let opctx = crate::context::op_context_for_saga_action(
+        &sagactx,
+        &params.serialized_authn,
+    );
+    let osagactx = sagactx.user_data();
+    let dpd_client: &dpd_client::Client = &osagactx.nexus().dpd_client;
+    let datastore = &osagactx.datastore();
+    let log = sagactx.user_data().log();
+
+    let (.., authz_instance, db_instance) = LookupPath::new(&opctx, &datastore)
+        .instance_id(instance_id)
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    let instance_id = db_instance.id();
+    let sled_uuid = db_instance.runtime_state.sled_id;
+
+    let (.., sled) = LookupPath::new(&osagactx.nexus().opctx_alloc, &datastore)
+        .sled_id(sled_uuid)
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    let sled_ip_address = sled.address();
+
+    debug!(log, "fetching network interfaces");
+
+    let network_interface = match datastore
+        .derive_guest_network_interface_info(&opctx, &authz_instance)
+        .await
+        .map_err(ActionError::action_failed)?
+        .into_iter()
+        .find(|interface| interface.primary)
+    {
+        Some(interface) => interface,
+        // Return early if instance does not have a primary network
+        // interface
+        None => return Ok(()),
+    };
+
+    let mac_address =
+        macaddr::MacAddr6::from_str(&network_interface.mac.to_string())
+            .map_err(|e| {
+                ActionError::action_failed(Error::internal_error(&format!(
+                    "failed to convert mac address: {e}"
+                )))
+            })?;
+
+    let vni: u32 = network_interface.vni.into();
+
+    debug!(log, "fetching external ip addresses");
+
+    let target_ip = &datastore
+        .instance_lookup_external_ips(&opctx, instance_id)
+        .await
+        .map_err(ActionError::action_failed)?
+        .get(which)
+        .ok_or_else(|| {
+            ActionError::action_failed(Error::internal_error(&format!(
+                "failed to find external ip address at index: {which}"
+            )))
+        })?
+        .to_owned();
+
+    debug!(log, "checking for existing nat mapping for {target_ip:#?}");
+
+    // TODO: currently if we have this environment variable set, we want to
+    // bypass all calls to DPD. This is mainly to facilitate some tests where
+    // we don't have dpd running. In the future we should probably have these
+    // testing environments running dpd-stub so that the full path can be tested.
+    if let Ok(_) = std::env::var("SKIP_ASIC_CONFIG") {
```

Comment:
Could we also print a debug message here? I worry we might have this set, but not realize it, during development, and wonder why our IP addresses aren't being forward to Dendrite.